README.md
# GreenHill README.md

## Description
GreenHill is a de novo chromosome-level scaffolding and phasing tool using Hi-C.
GreenHill generates chromosome-level haplotypes by scaffolding and phasing
the input contigs using a combination of information from Hi-C and other reads (PE, MP, LongRead).

## Version
v1.0.0

## Author
Shun Ouchi and Rei Kajitani at Tokyo Institute of Technology wrote key source codes.
Address for this tool: <platanus@bio.titech.ac.jp>


## Requirements
* GCC
    - <https://gcc.gnu.org/>
    - version >= 4.4, with OpenMP
    - To compile the source code.

* Minimap2
    - <https://github.com/lh3/minimap2>
    - Only required to use PacBio/Oxford-Nanopore long reads.

## Installation
### Install from source
Install the dependencies above.
* GCC
* Minimap2

Compile (make), and copy greenhill to a directory listed in PATH.
```sh
git clone https://github.com/ShunOuchi/GreenHill.git
cd src
make
cp greenhill <installation_path>
```
### Install through conda
On a new conda environment, run:
```sh
conda install -c bioconda greenhill
```
Or, if you want to have a separate conda environment for GreenHill
```sh
conda create -n greenhill -c bioconda greenhill
```
You will then need to activate the greenhill envirionment before using it with:
```sh
conda activate greenhill
```


## Synopsis
### Inputs
* Input assembly (required)
    * Haplotype-aware style input (such as [Platanus-allee](http://platanus.bio.titech.ac.jp/platanus2)):<br>
      primaryBubble.fa secondaryBubble.fa nonBubble.fa
    * Pseudo-haplotype style input (such as [FALCON-Unzip](https://github.com/PacificBiosciences/FALCON_unzip)) or Mixed-haplotype style input (such as [Canu](https://github.com/marbl/canu)):<br>
      contigs.fa
* Input reads
    * Hi-C reads: HIC_1.fq HIC_2.fq (required)
    * Illumina paired-end: PE_1.fq PE_2.fq (optional)
    * Illumina mate-pair : MP_1.fq MP_2.fq (optional)
    * PacBio/Oxford-Nanopore long reads  : longread.fq (optional)

### Commands

* for Haplotype-aware style input
```
greenhill \
-c nonBubble.fa \
-b primaryBubble.fa secondaryBubble.fa \
-IP1 PE_1.fq PE_2.fq \
-OP2 MP_1.fq MP_2.fq \
-p longread.fq \
-HIC HIC_1.fq HIC_2.fq \
2>3D.log
```

* for Pseudo-haplotype or Mixed-haplotype style input
```
greenhill \
-cph contigs.fa \
-IP1 PE_1.fq PE_2.fq \
-OP2 MP_1.fq MP_2.fq \
-p longread.fq \
-HIC HIC_1.fq HIC_2.fq \
2>3D.log
```

### Final output
    out_afterPhase.fa (phased diploid scaffolds)


---
## Example
Below is showing examples how to run GreenHill using test dataset.
The test dataset is the simulated diploid dataset of _Caenorhabditis elegans_ chr1.
### Example 1. I have Platanus-allee assembly (Haplotype-aware style input)
```
greenhill \
-c Platanus-allee_result/out_nonBubbleOther.fa \
-b Platanus-allee_result/out_primaryBubble.fa Platanus-allee_result/out_secondaryBubble.fa \
-IP1 reads/PE_*.fq.gz \
-OP2 reads/MP5k_*.fq.gz \
-OP3 reads/MP9k_*.fq.gz \
-p reads/longread.fq.gz \
-HIC reads/HIC_1.fq.gz reads/HIC_2.fq.gz
```
### Example 2. I have FALCON-Unzip assembly (Psuedo-haplotype style input)
```
greenhill \
-cph FALCON-Unzip_result/cns_p_ctg.fa FALCON-Unzip_result/cns_h_ctg.fa \
-p reads/longread.fq.gz \
-HIC reads/HIC_1.fq.gz reads/HIC_2.fq.gz
```
### Example 3. I have Canu assembly (Mixed-haplotype style input)
```
greenhill \
-cph Canu_result/asm.contigs.fa \
-p reads/longread.fq.gz \
-HIC reads/HIC_1.fq.gz reads/HIC_2.fq.gz
```

---
## Usage
### Command
```sh
greenhill [OPTIONS] 2>log
```
### Options
    -o STR                             : prefix of output file and directory (do not use "/", default out, length <= 200)
    -c FILE1 [FILE2 ...]               : contig (or scaffold) file (fasta format; for Haplotype-aware style input)
    -b FILE1 [FILE2 ...]               : bubble seq file (fasta format; for Haplotype-aware style input)
    -cph FILE1 [FILE2 ...]             : contig (or scaffold) file (fasta format; for Pseudo-haplotype or Mixed-haplotype style input; only effective without -c, -b option)
    -ip{INT} PAIR1 [PAIR2 ...]         : lib_id inward_pair_file (interleaved file, fasta or fastq)
    -IP{INT} FWD1 REV1 [FWD2 REV2 ...] : lib_id inward_pair_files (separate forward and reverse files, fasta or fastq)
    -op{INT} PAIR1 [PAIR2 ...]         : lib_id outward_pair_file (interleaved, fasta or fastq)
    -OP{INT} FWD1 REV1 [FWD2 REV2 ...] : lib_id outward_pair_files (separate forward and reverse files, fasta or fastq)
    -p PAIR1 [PAIR2 ...]               : long-read file (PacBio, Nanopore) (fasta or fastq)
    -hic PAIR1 [PAIR2 ...]             : HiC_pair_files (reads in 1 file, fasta or fastq)
    -HIC FWD1 REV1 [FWD2 REV2 ...]     : HiC_pair_files (reads in 2 files, fasta or fastq)
    -t INT                             : number of threads (default 1)
    -tmp DIR                           : directory for temporary files (default .)
    -l INT                             : minimum number of links to scaffold (default 3)
    -k INT                             : minimum number of links to phase variants (default 1)
    -s INT1 [INT2 ...]                 : mapping seed length for short reads (default 32 64 96)
    -mapper FILE                       : path of mapper executable file (default minimap2, only effective with -p option)
    -minimap2_sensitive                : sensitive mode for minimap2 (default, off; only effective with -p option)

### Input format:
   Uncompressed and compressed (gzip or bzip2) files are accepted for -c, -ip, -IP, -op, -OP, -p, -hic and -HIC option.

### Final output:
   PREFIX_afterPhase.fa

### Other misc outputs:
   PREFIX_*

PREFIX is specified by -o


---
## Notes
* Compressed input files

Both uncompressed and compressed (gzip or bzip2) FASTA/FASTQ files are accepted.
Formats are auto-detected. Internally, "file -bL", "gzip -cd" and "bzip2 -cd" commands, which can be
used in most of the UNIX OSs, are utilized.

* Minimap2

This tool is used to align PacBio/Oxford-Nanopore long reads.
When long reads are input through the -p option, please check Minimap2 is installed as "minimap2" command
or specify the path of Minimap2 using the -mapper option.

* Paired-end (mate-pair) input

Paired libraries are classified into "inward-pair" and "outward-pair" according to the sequence direction.
For file formats, separate and interleaved files can be input through -IP (-OP) and -ip (-op)
options, respectively.

Inward-pair (usually called "paired-end", accepted in options "-IP" or "-ip"):

    FWD --->
        5' -------------------- 3'
        3' -------------------- 5'
                        <--- REV

Outward-pair (usually called "mate-pair", accepted in options "-OP" or "-op"):

                        ---> REV
        5' -------------------- 3'
        3' -------------------- 5'
    FWD <---

Example inputs:

    Inward-pair (separate, insert=300)   : PE300_1.fq PE300_2.fq
    Inward-pair (interleaved, insert=500): PE500_pair.fq
    Outward-pair (separate, insert=2k)   : MP2k_1.fa MP2k_2.fq

Corresponding options:

    -IP1 PE300_1_pair.fq PE300_2.fq \
    -ip2 PE500_pair.fq \
    -OP3 MP2k_1.fq MP2k_2.fq



Platanus-allee_v2.2.2_modified/ChangeLog.md
# Platanus-allee ChangeLog.md

## Version 2.2.2 (17 May 2019)
* To reduce false-detection of heterozygous regions, the fork-anchor function is deprecated.
* To reduce mis-assemblies, the partial-gap-close function in the phasing step is deprecated.
* Add -no_partial option of the consensus command to turn off the partial-gap-close function.

## Version 2.2.1 (8 May 2019)
* Improve performance to utilize long-reads.
* Add aggressive mode for phase and consensus commands (-aggressive). It is experimental and not recommended. 
* Simplify output-files of consensus command.

## Version 2.2.0 (25 Apr. 2019)
* Rename and simplify output file-names.
* Add gap-closing function in the "consensus" command.
* Improve ability to detect homologous haplotype-pairs.
* Reduce mis-assemblies for many cases.
* Reduce gap-rates using partial-gap-close function.
* Improve the speed of read-mapping.

## Version 2.1.0 (20 Nov. 2018)
* Add "divide" command and functions to divide erroneous sequences in a phasing step.

## Version 2.0.2 (16 Sep. 2018)
* Fixed a bug in the untangling function for cross-structures in graphs (phase, solve_DBG commands).
* Compressed (gzip or bzip2) FASTA/FASTQ files can be accepted.

## Version 2.0.1 (17 Aug. 2018)
* Fixed a bug causing the segmentation fault in gap_close.
* The result is the same as that of Version 2.0.0 if it finishes a process.

## Version 2.0.0 (9 Jul. 2018)
* The version used to write the manuscript (under review).
* Initial release as an exutable file for Linux x86_64.


phase.cpp
/*
Copyright (C) 2018 Itoh Laboratory, Tokyo Institute of Technology

This file is part of Platanus-allee.

Platanus-allee is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

Platanus-allee is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with Platanus-allee; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "phase.h"
#include <sys/stat.h>
#include <sys/types.h>

using std::vector;
using std::string;


//////////////////////////////////////////////////////////////////////////////////////
// constructor
//////////////////////////////////////////////////////////////////////////////////////
Phase::Phase()
{
    optionSingleArgs["-o"] = "out";
    optionSingleArgs["-e"] = "";
    optionSingleArgs["-v"] = "32";
    optionSingleArgs["-k"] = "1";
    optionSingleArgs["-l"] = "3";
    optionSingleArgs["-L"] = "200000";
    optionSingleArgs["-t"] = "1";
    optionMultiArgs["-c"] = vector<string>();
    optionSingleArgs["-i"] = "2";
    optionSingleArgs["-tmp"] = ".";

    optionMultiArgs["-s"] = vector<string>(3);
    optionMultiArgs["-s"][0] = "32";
    optionMultiArgs["-s"][1] = "64";
    optionMultiArgs["-s"][2] = "96";

    pairedEndSingleFileType.push_back("-ip");
    pairedEndSingleFileType.push_back("-op");
    pairedEndPairFileType.push_back("-IP");
    pairedEndPairFileType.push_back("-OP");
    optionMultiArgs["-p"] = vector<string>();
    optionMultiArgs["-x"] = vector<string>();
    optionMultiArgs["-X"] = vector<string>();
    optionMultiArgs["-hic"] = vector<string>(); //added by ouchi
    optionMultiArgs["-HIC"] = vector<string>(); //added by ouchi

    optionSingleArgs["-mapper"] = "";
    optionBool["-minimap2_sensitive"] = false;
    optionBool["-minimap2"] = false;
    optionBool["-minialign"] = false;
    optionBool["-kmer_align"] = false;

    optionBool["-reduce_redundancy"] = true;

    optionBool["-aggressive"] = false;
    optionBool["-no_partial"] = false;
}




//////////////////////////////////////////////////////////////////////////////////////
// usage
//////////////////////////////////////////////////////////////////////////////////////
void Phase::usage(void) const
{

    std::cerr << "\nUsage: platanus_allee phase [Options]\n"
              << "Options:\n"
              << "    -o STR                             : prefix of output file and directory (do not use \"/\", default " << optionSingleArgs.at("-o") << ", length <= " << platanus::ConstParam::MAX_FILE_LEN << ")\n"
              << "    -c FILE1 [FILE2 ...]               : contig (or scaffold) file (fasta format)\n"
              << "    -ip{INT} PAIR1 [PAIR2 ...]         : lib_id inward_pair_file (interleaved file, fasta or fastq)\n"
              << "    -IP{INT} FWD1 REV1 [FWD2 REV2 ...] : lib_id inward_pair_files (separate forward and reverse files, fasta or fastq)\n"
              << "    -op{INT} PAIR1 [PAIR2 ...]         : lib_id outward_pair_file (interleaved, fasta or fastq)\n"
              << "    -OP{INT} FWD1 REV1 [FWD2 REV2 ...] : lib_id outward_pair_files (separate forward and reverse files, fasta or fastq)\n"
              << "    -p PAIR1 [PAIR2 ...]               : long-read file (PacBio, Nanopore) (fasta or fastq)\n"
              << "    -x PAIR1 [PAIR2 ...]               : linked-reads files (paired-ends, 10x Genomics) (interleaved, fasta or fastq)\n"
              << "    -X FWD1 REV1 [FWD2 REV2 ...]       : linked-reads files (paired-ends, 10x Genomics) (separate forward and reverse files, fasta or fastq)\n"
              << "    -hic PAIR1 [PAIR2 ...]             : HiC-reads files (paired-ends) (interleaved, fasta or fastq)\n" //added by ouchi
              << "    -HIC FWD1 REV1 [FWD2 REV2 ...]     : HiC-reads files (paired-ends) (separate forward and reverse files, fasta or fastq)\n" //added by ouchi
              << "    -i INT                             : number of iterations (default " << optionSingleArgs.at("-i") << ")\n"
              << "    -l INT                             : minimum number of links to scaffold (default " << optionSingleArgs.at("-l") << ")\n"
              << "    -k INT                             : minimum number of links to phase variants (default " << optionSingleArgs.at("-k") << ")\n"
              << "    -s INT1 [INT2 ...]                 : mapping seed length for short reads (default " << optionMultiArgs.at("-s")[0] << " " << optionMultiArgs.at("-s")[1] << " " << optionMultiArgs.at("-s")[2]  << ")\n"
              << "    -t INT                             : number of threads (default " << optionSingleArgs.at("-t") << ")\n"
              << "    -tmp DIR                           : directory for temporary files (default " << optionSingleArgs.at("-tmp") << ")\n"
              << "    -mapper FILE                       : path of mapper executable file (default minimap2, only effective with -p option)\n"
              << "    -minimap2_sensitive                : sensitive mode for minimap2 (default, off; only effective with -p option)\n"
              << "    -aggressive                        : aggressively extend scaffolds (default off)\n"
//              << "    -no_partial                        : not close gaps partially, i.e. only close ones completely (default, off)\n"
//              << "    -reduce_redundancy                 : reduce redundant sequences that exactly matche others (default, off)\n"
//              << "    -minialign                         : use minialign insterd of minimap2 (default) for alignment of long reads\n"
//              << "    -kmer_align                        : use built-in alignment based on k-mer match insterd of minimap (default) for long reads\n"
              << "\n\n"

              << "Input format:\n"
              << "    Uncompressed and compressed (gzip or bzip2) files are accepted for -c, -ip, -IP, -op, -OP, -p, -x, -X, -hic and -HIC options.\n" //edited by ouchi
              << "\n\n"

              << "Outputs:\n"
              << "    PREFIX_allPhasedScaffold.fa (including sequences below)\n"
              << "    PREFIX_primaryBubble.fa\n"
              << "    PREFIX_secondaryBubble.fa\n"
              << "    PREFIX_nonBubbleHetero.fa\n"
              << "    PREFIX_nonBubbleOther.fa\n"
              << "    PREFIX_consensusInput.fa (for \"consensus\" command (-c))\n"
              << "\n"
              << "Note, PREFIX is specified by -o\n"
              << std::endl;
}


void Phase::exec(void)
{
	platanus::setGlobalTmpFileDir(optionSingleArgs["-tmp"].c_str());

	string intermediateDirectoryName = optionSingleArgs["-o"];
	intermediateDirectoryName += "_intermediateResults";
	this->setDirectoryName(intermediateDirectoryName);
	this->createDirectory();

    unsigned long numIterate = atoi(optionSingleArgs["-i"].c_str());
    if (!(optionMultiArgs["-p"].empty())) {
		++numIterate;
	}

    for (unsigned long times = 1; times <= numIterate; ++times) {
		std::ostringstream oss;
		oss << intermediateDirectoryName << "/round" << times;
		this->setDirectoryName(oss.str());

        this->createDirectory();
		this->execSolveDBG(times, numIterate);
		this->execGapClose(times, numIterate);
    }

	this->moveAndConcatenateFinalRoundResult(intermediateDirectoryName, numIterate);

	std::cerr << "phase completed!" << std::endl;
}


void Phase::setDirectoryName(const string newDirectoryName)
{
	this->previousDirectoryName = this->directoryName;
	this->directoryName = newDirectoryName;
}


void Phase::createDirectory(void) const
{
    struct stat st;

    if (stat(directoryName.c_str(), &st) != 0) {
        int returnValue = mkdir(directoryName.c_str(), 0755);
        if (returnValue != 0) {
            throw platanus::CreateDirError(directoryName);
        }
    }
}


void Phase::execSolveDBG(const unsigned long times, const unsigned long numIterate)
{
	vector<string> boolArgOptionList = {"-minimap2_sensitive", "-aggressive", "-reduce_redundancy"};
	vector<string> singleArgOptionList = {"-t", "-tmp", "-l", "-k", "-mapper"};
	vector<string> multiArgOptionList = {"-x", "-X", "-hic", "-HIC", "-s"}; //edited by ouchi
	if (times > 1) {
		multiArgOptionList.push_back("-p");
	}

	string inputPrefix = this->previousDirectoryName;
	inputPrefix += "/";
	inputPrefix += optionSingleArgs["-o"];

	string outputPrefix = this->directoryName;
	outputPrefix += "/";
	outputPrefix += optionSingleArgs["-o"];


    std::ostringstream oss;
    oss << this->platanusBinary << " solve_DBG";

	if (times == 1) {
		multiArgOptionList.push_back("-c");
		multiArgOptionList.push_back("-b");
	}
	else {
        oss << " -c " << inputPrefix << "_gapClosed_nonBubbleOther.fa"
			<< " -b " << inputPrefix << "_gapClosed_primaryBubble.fa" << " " << inputPrefix << "_gapClosed_secondaryBubble.fa" << " " << inputPrefix << "_gapClosed_primaryFork.fa" << " " << inputPrefix << "_gapClosed_secondaryFork.fa";
	}

	for (auto optItr = boolArgOptionList.begin(); optItr != boolArgOptionList.end(); ++optItr) {
		if (optionBool[*optItr])
			oss << " " << *optItr;
	}

    for (auto itr = optionPairFile.begin(); itr != optionPairFile.end(); ++itr) {
        oss << " " << itr->libraryType << itr->libraryID << " " << itr->fileFirst;
        if (!(itr->fileSecond.empty())) {
            oss << " " << itr->fileSecond;
        }
    }

	for (auto optItr = singleArgOptionList.begin(); optItr != singleArgOptionList.end(); ++optItr) {
		if (!(optionSingleArgs[*optItr].empty())) {
			oss << " " << *optItr << " " << optionSingleArgs[*optItr];
		}
	}

	for (auto optItr = multiArgOptionList.begin(); optItr != multiArgOptionList.end(); ++optItr) {
		if (!(optionMultiArgs[*optItr].empty())) {
			oss << " " << *optItr;
			for (auto argItr = optionMultiArgs[*optItr].begin(); argItr != optionMultiArgs[*optItr].end(); ++argItr) {
				oss << " " << *argItr;
			}
		}
	}

	oss << " -o " << outputPrefix;
    oss << " 2>" << outputPrefix << ".solveDBGLog";

    if (system(oss.str().c_str()) != 0) {
        throw platanus::SolveDBGError();
    }
}


void Phase::execGapClose(const unsigned long times, const unsigned long numIterate)
{
	string inputPrefix = this->directoryName;
	inputPrefix += "/";
	inputPrefix += optionSingleArgs["-o"];

	string outputPrefix = this->directoryName;
	outputPrefix += "/";
	outputPrefix += optionSingleArgs["-o"];

	vector<string> singleArgOptionList = {"-t", "-tmp"};
	vector<string> multiArgOptionList = {"-s"};
	vector<string> boolArgOptionList = {"-no_partial", "-reduce_redundancy"};

    std::ostringstream oss;
    oss << this->platanusBinary << " gap_close";

	oss << " -c"
		<< " " << inputPrefix << "_primaryBubble.fa"
		<< " " << inputPrefix << "_secondaryBubble.fa"
		<< " " << inputPrefix << "_primaryFork.fa"
		<< " " << inputPrefix << "_secondaryFork.fa"
		<< " " << inputPrefix << "_nestedBubble.fa"
		<< " " << inputPrefix << "_nonBubbleOther.fa";

    for (auto itr = optionPairFile.begin(); itr != optionPairFile.end(); ++itr) {
        oss << " " << itr->libraryType << itr->libraryID << " " << itr->fileFirst;
        if (!(itr->fileSecond.empty())) {
            oss << " " << itr->fileSecond;
        }
    }

	for (auto optItr = singleArgOptionList.begin(); optItr != singleArgOptionList.end(); ++optItr) {
		if (!(optionSingleArgs[*optItr].empty())) {
			oss << " " << *optItr << " " << optionSingleArgs[*optItr];
		}
	}

	for (auto optItr = multiArgOptionList.begin(); optItr != multiArgOptionList.end(); ++optItr) {
		if (!(optionMultiArgs[*optItr].empty())) {
			oss << " " << *optItr;
			for (auto argItr = optionMultiArgs[*optItr].begin(); argItr != optionMultiArgs[*optItr].end(); ++argItr) {
				oss << " " << *argItr;
			}
		}
	}

	for (auto optItr = boolArgOptionList.begin(); optItr != boolArgOptionList.end(); ++optItr) {
		if (optionBool[*optItr])
			oss << " " << *optItr;
	}

	oss << " -o " << outputPrefix;
    oss << " 2>" << outputPrefix << ".gapCloseLog";

    if (system(oss.str().c_str()) != 0) {
        throw platanus::GapError();
    }
}

void Phase::moveAndConcatenateFinalRoundResult(const string intermediateDirectoryName, const unsigned long times)
{
    std::ostringstream prefixOss;
	prefixOss << intermediateDirectoryName << "/round" << times << "/" << optionSingleArgs["-o"];
	string finalRoundPrefix = prefixOss.str();

    std::ostringstream cmdOss;
    cmdOss << "mv " << finalRoundPrefix << "_gapClosed_primaryBubble.fa" << " " << optionSingleArgs["-o"] << "_primaryBubble.fa" << "; "
		<< "mv " << finalRoundPrefix << "_gapClosed_secondaryBubble.fa" << " " << optionSingleArgs["-o"] << "_secondaryBubble.fa" << "; "
		<< "mv " << finalRoundPrefix << "_gapClosed_nonBubbleOther.fa" << " " << optionSingleArgs["-o"] << "_nonBubbleOther.fa" << "; "
		<< "cat " << optionSingleArgs["-o"] << "_primaryBubble.fa" << " "
		<< optionSingleArgs["-o"] << "_secondaryBubble.fa" << " "
		<< finalRoundPrefix << "_gapClosed_primaryFork.fa" << " "
		<< finalRoundPrefix << "_gapClosed_secondaryFork.fa" << " "
		<< finalRoundPrefix << "_gapClosed_nestedBubble.fa" << " "
		<< optionSingleArgs["-o"] << "_nonBubbleOther.fa" << " "
		<< ">" << optionSingleArgs["-o"] <<  "_allPhasedScaffold.fa" << "; "

		<< "cat " << finalRoundPrefix << "_gapClosed_primaryFork.fa" << " "
		<< finalRoundPrefix << "_gapClosed_secondaryFork.fa" << " "
		<< finalRoundPrefix << "_gapClosed_nestedBubble.fa" << " "
		<< ">" << optionSingleArgs["-o"] <<  "_nonBubbleHetero.fa" << "; "

		<< "cat " << optionSingleArgs["-o"] << "_primaryBubble.fa" << " "
		<< finalRoundPrefix << "_gapClosed_primaryFork.fa" << " "
		<< optionSingleArgs["-o"] << "_nonBubbleOther.fa" << " "
		<< ">" << optionSingleArgs["-o"] <<  "_consensusInput.fa";

    if (system(cmdOss.str().c_str()) != 0) {
        throw platanus::CreateLinkError();
    }
}
